### PR TITLE
Caret position in fourslash

### LINKED ISSUE #16877: caret positioning in fourslash
**TypeScript Version:**  master, commit

**Code**

The following fourslash test fails:
```ts
/// <reference path="fourslash.ts"/>

////var /**/[a,b,

goTo.marker("");
debug.printCurrentFileState();
edit.insertLine("");
debug.printCurrentFileState();
edit.insertLine("");
debug.printCurrentFileState();
verify.currentFileContentIs(`var

    [a, b,`);
```

**Expected behavior:**
test succeeds
**Actual behavior:**
test fails because caret is incremented too far because of the formatting edit applied to add a space between `,` and `b` after the first newline. The second newline is placed after `[`.

### PR DESCRIPTION
**Pull request by aozgaa**
Fixes #16877.

### CODE REVIEW

File: tests/cases/fourslash/smartIndentObjectBindingPattern01.ts [hunk 1]
```diff
@@ -1,14 +1,16 @@
...
- 
+
 function verifyIndentationAfterNewLine(marker: string, indentation: number): void {
     goTo.marker(marker);
     edit.insert("\r\n");
     verify.indentationIs(indentation);
 }
 
-verifyIndentationAfterNewLine("1", 4);
+// TODO (arozga): fix this.
```
Comment (RyanCavanaugh) [inline tests/cases/fourslash/smartIndentObjectBindingPattern01.ts]: Is fixed now?

File: src/harness/fourslash.ts [hunk 2]
```diff
@@ -1740,31 +1722,25 @@ namespace FourSlash {
                 }
```
Comment (andy-ms) [inline src/harness/fourslash.ts]: Do you know why above in `if (tyhis.enableFormatting)` we update `offset` but not `currentCaretPosition`?
Also, is `fixCaretPosition` still necessary if we update it correctly?

File: src/harness/fourslash.ts [hunk 3]
```diff
@@ -1804,19 +1780,34 @@ namespace FourSlash {
...
+         * May be negative.
+         */
         private applyEdits(fileName: string, edits: ts.TextChange[], isFormattingEdit = false): number {
             // We get back a set of edits, but langSvc.editScript only accepts one at a time. Use this to keep track
-            // of the incremental offset from each edit to the next. Assumption is that these edit ranges don't overlap
-            let runningOffset = 0;
+            // of the incremental offset from each edit to the next. We assume these edit ranges don't overlap
+
             edits = edits.sort((a, b) => a.span.start - b.span.start);
+            for (let i = 0; i < edits.length - 1; ++i) {
```
Comment (andy-ms) [inline src/harness/fourslash.ts]: Nit: `i++` is more conventional

File: src/harness/fourslash.ts [hunk 4]
```diff
@@ -1804,19 +1780,34 @@ namespace FourSlash {
             }
         }
 
+        /**
+         * @returns The number of characters added to the file as a result of the edits.
+         * May be negative.
+         */
         private applyEdits(fileName: string, edits: ts.TextChange[], isFormattingEdit = false): number {
```
Comment (andy-ms) [inline src/harness/fourslash.ts]: The `isFormattingEdit` parameter is omitted only once; consider making it non-optional. Same for the parameters in `printCurrentFileState`.

File: tests/cases/fourslash/smartIndentObjectBindingPattern01.ts (hunk 1 above)
Comment (aozgaa) [inline tests/cases/fourslash/smartIndentObjectBindingPattern01.ts]: No. This problem actually already existed, but only by adjusting the caret correctly is the issue exposed. But actually fixing indenting behaviors is outside the scope of this PR. When this change get's accepted, I'll make a new issue for this.

File: src/harness/fourslash.ts (hunk 2 above)
Comment (aozgaa) [inline src/harness/fourslash.ts]: `this.applyEdits` returns the number of characters added/removed in the file, but this is different from the adjustment to the caret position (which is the number of characters added/removed in the file *before the caret*).

Comment (aozgaa) [inline src/harness/fourslash.ts]: `fixCaretPosition` appears to be unnecessary now.

File: src/harness/fourslash.ts (hunk 4 above)
Comment (aozgaa) [inline src/harness/fourslash.ts]: changed.

File: src/harness/fourslash.ts (hunk 3 above)
Comment (aozgaa) [inline src/harness/fourslash.ts]: changed.

**Pull request merged**